cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/PluginClassLoader.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.adapter;
+
+import co.cask.cdap.app.program.ManifestFields;
+import co.cask.cdap.common.lang.CombineClassLoader;
+import co.cask.cdap.common.lang.DirectoryClassLoader;
+import co.cask.cdap.common.lang.PackageFilterClassLoader;
+import co.cask.cdap.common.lang.ProgramClassLoader;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Predicates;
+import com.google.common.collect.ImmutableList;
+
+import java.io.File;
+import java.util.Set;
+import java.util.jar.Manifest;
+
+/**
+ * ClassLoader for template plugin. The ClassLoader hierarchy is pretty complicated.
+ * <p/>
+ * First, we have the "Plugin Lib ClassLoader".
+ *
+ * <pre>{@code
+ *            CDAP System CL
+ *                  ^
+ *                  |
+ *          Program Filter CL (cdap-api and hadoop classes only)
+ *                  ^
+ *                  |
+ *             Template CL (expanded app bundle jar)
+ *                  ^
+ *                  |
+ *          Template Filter CL (Export-Package classes only)
+ *                  ^
+ *                  |
+ *           Plugin Lib CL (Common library for plugin)
+ * }</pre>
+ *
+ * <p/>
+ * Then, the parent of the PluginClassLoader is formed by a {@link CombineClassLoader} of
+ * {@code (Program Filter CL, Plugin Lib ClassLoader) }. It is a combine class loader because we
+ * want the cdap-api classes not affected by the "Template Filter ClassLoader" (which usually would have cdap-api
+ * classes filtered out).
+ *
```

Comment:
Nice documentation, this makes it really clear. Are there any test cases for this class loader?